pkg/kv/kvserver/kvserverbase/nondeterministic_failure.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+//
+
+package kvserverbase
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
+)
+
+// NonDeterministicFailureError is an error type that indicates that a state machine
+// transition failed due to an unexpected error. Failure to perform a state
+// transition is a form of non-determinism, so it can't be permitted for any
+// reason during the application phase of state machine replication. The only
+// acceptable recourse is to signal that the replica has become corrupted.
+//
+// All errors returned by replicaDecoder and replicaStateMachine will be instances
+// of this type.
+type NonDeterministicFailureError struct {
+	wrapped error
+	expl    redact.RedactableString
```

Comment:
Ugh, I had this part which isn't a pure move in a separate commit but seem to have accidentally merged them.
This field used to be a `string` which was playing fast and loose with redaction. Now it's a redactable string (and this is also now the return type of `GetNonDeterministicFailureExplanation`, after redaction.
Previously we `fmt.Sprintf`'ed into the string, now we `redact.Sprintf`. And the callers to `GetNonDeterministicFailureExplanation(err)` all turn it into a string via `Redact()`, i.e. replace unsafe bits by `<x>`. So we've made sure no PII will leak into the "non-sensitive cue" (second return arg):

https://github.com/cockroachdb/cockroach/blob/db43c8f5a3798efaf0db804af623140c6d4bdc95/pkg/kv/kvserver/replica_raft.go#L679-L683